File: 41. First Missing Positive.py
class Solution(object):
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 1
        
        n = len(nums)
        i = 0
        ret = n+1
        while i < n:
            e = nums[i]
            if i+1 != e and 1<= e <= n and nums[e-1] != e:
                nums[i], nums[e-1] = nums[e-1], e
            else:
                i += 1
        for i, e in enumerate(nums):
            if i+1 != e:
                return i+1
        return ret

File: 41. First Missing Positive.go
func firstMissingPositive(nums []int) int {
	n := len(nums)
	if n == 0 {
		return 1
	}
	swap := func(i, j int) {
		nums[i], nums[j] = nums[j], nums[i]
	}
	for i := 0; i < n; i++ {
		for nums[i] > 0 && nums[i] <= n && i+1 != nums[i] && nums[i] != nums[nums[i]-1] {
			swap(i, nums[i]-1)
		}
	}
	for i := 0; i < n; i++ {
		if i+1 != nums[i] {
			return i + 1
		}
	}
	return nums[n-1] + 1
}